[pack] Add ability to register functions from extensions V2

ankitkumarr opened the pull request:
Did a quick check by having an extension implement and register `IFunctionProvider`. It worked as expected.
Fixes #5596

ankitkumarr commented on src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs:
```diff
@@ -61,6 +61,7 @@ public Type[] GetStartupTypes()
+            // TODO: take a look at this
             string binPath;
             var functionMetadataCollection = _functionMetadataProvider.GetFunctionMetadata(forceRefresh: true);
             HashSet<string> bindingsSet = null;
```
It looks like, when extension bundle is configured, we first look at the functions to understand which startup types to load. However, this would be a chicken and egg problem if extensions themselves are indexing the functions.

This is ok if the extension using custom indexer is not part of extension bundle. But when they are, this part may get tricky.

ankitkumarr commented on src/WebJobs.Script.Abstractions/WebJobs.Script.Abstractions.csproj:
```diff
@@ -0,0 +1,20 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <Import Project="..\..\build\common.props" />
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
```
Need to add property to use "Microsoft.Azure.WebJobs.Script.Abstractions" for dll name

ankitkumarr commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -49,6 +51,14 @@ internal ImmutableArray<FunctionMetadata> LoadFunctionMetadata()
+            if (_indexers.Count() > 0)
+            {
+                functionMetadataList.AddRange(GetMetadataFromCustomIndexers());
+                _functionErrors.AddRange(GetErrorsFromCustomIndexers());
```
Do we want to worry about duplicates? For now, I didn't.

ankitkumarr commented on src/WebJobs.Script.Abstractions/Description/IFunctionProvider.cs:
```diff
@@ -0,0 +1,26 @@
+        /// Gets all function metadata that this provider knows about
+        /// </summary>
+        /// <returns>An IEnumerable of FunctionMetadata</returns>
+        IEnumerable<FunctionMetadata> GetFunctionMetadata();
```
It was requested that this be made `async` (though this is going to be a blocking call) so in future, extensions can take advantage of async if that happens, without restructuring their code.

mathewc replied:
Yes, making this async would be good

mathewc commented on src/WebJobs.Script.Abstractions/WebJobs.Script.Abstractions.csproj:
```diff
@@ -0,0 +1,26 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <Import Project="..\..\build\common.props" />
+    <PropertyGroup>
```
The files that you moved to this assembly - are you going to change their namespaces to Microsoft.Azure.WebJobs.Script.Abstractions?

ankitkumarr replied:
I started changing them, but didn't end up going through to avoid issues and a lot of refactoring. Would you suggest changing the namespaces for clarity?

mathewc replied:
I think the important thing is to have the new package types in the proper public namespace going forward. I realize that'll create churn but it'd be good to get the folder structure + naming for the new package clean.

ankitkumarr replied:
Fair! I updated the PR with the proper changes. Once this is merged in, I will update Core Tools to also change the namespace. I don't think we have any other dependency that would need an update.

mathewc commented on src/WebJobs.Script/Description/DotNet/DependencyHelper.cs:
```diff
@@ -165,5 +166,33 @@ public static List<string> GetRuntimeFallbacks(string rid)
+        /// </summary>
+        /// <param name="assemblyFormatString"> string representing assembly information</param>
+        /// <returns> bool if string in shared assembly representation format. </returns>
+        public static bool IsSharedAssemblyFormat(string assemblyFormatString)
```
Add unit tests for this and all other new methods. There's already a DendencyHelperTests suite

mathewc commented on src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs:
```diff
@@ -78,6 +78,11 @@ public sealed class DotNetFunctionInvoker : FunctionInvokerBase
         {
+            if (DependencyHelper.IsSharedAssemblyFormat(functionMetadata.ScriptFile))
+            {
+                return new NoOpFunctionMetadataResolver();
```
add some comments here describing what's going on

mathewc commented on src/WebJobs.Script/Description/DotNet/NoOpFunctionMetadataResolver.cs:
```diff
@@ -0,0 +1,56 @@
+
+        public IReadOnlyCollection<string> GetCompilationReferences()
+        {
+            return new ReadOnlyCollection<string>(new List<string>());
```
Don't new these up each time. Definitely you don't need to new up the inner list. Can't you return something derived from Enumerable.Empty<string>?

ankitkumarr replied:
ahhh right right! Yeah that makes sense

mathewc commented on src/WebJobs.Script/Description/DotNet/NoOpFunctionMetadataResolver.cs:
```diff
@@ -0,0 +1,56 @@
+
+        public Task<PackageRestoreResult> RestorePackagesAsync()
+        {
+            return Task.FromResult(new PackageRestoreResult()
```
return a lazy cached object here rather than new up each time

mathewc commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -22,16 +23,18 @@ public class FunctionMetadataManager : IFunctionMetadataManager
         private Dictionary<string, ICollection<string>> _functionErrors = new Dictionary<string, ICollection<string>>();
 
-        public FunctionMetadataManager(IOptions<ScriptJobHostOptions> scriptOptions, IFunctionMetadataProvider functionMetadataProvider, IOptions<HttpWorkerOptions> httpWorkerOptions, ILoggerFactory loggerFactory)
+        public FunctionMetadataManager(IOptions<ScriptJobHostOptions> scriptOptions, IFunctionMetadataProvider functionMetadataProvider, IEnumerable<IFunctionProvider> functionProviders, IOptions<HttpWorkerOptions> httpWorkerOptions, ILoggerFactory loggerFactory)
```
I think we should be unifying ProxyMetadataManager and your new providers. Consulting the custom providers should be done inside the FunctionMetadataProvider - it should handle proxies and these new custom providers. ProxyMetadataManager should implement your new interface - it already implements the same contract. This way we prove out the extensibility model, which we're already using internally for proxies.

ankitkumarr replied:
I am trying to think through how would that look like. `ProxyMetadaManager` doesn't implement the same contract as `IFunctionMetadataProvider` today. It uses different proxy types like `ProxyMetadaInfo`, which has a proxy specific executor, in it.

Having it converged would mean, we modify the `IFunctionMetadataProvider` contract and change the shape of the metadata somehow.

Unless there's something I am missing? Did you have a specific way in mind to deal with proxy specific metdata, proxy executor, proxy descriptors etc?

mathewc replied:
It is an IFunctionsMetadataProvider indirectly in that it fulfills the same contract via ProxyMetadataInfo. ScriptHost is accessing this at the same point that it's querying the rest via FunctionMetadataManager.Functions. All the proxy metadata should come through that one call [here](https://github.com/Azure/azure-functions-host/blob/f8fb1efb36e288c38c283af0c9c81b8b91f7d46f/src/WebJobs.Script/Host/ScriptHost.cs#L338). The proxies metadata provider would return ProxyFunctionMetadata types that subclass just to add the ProxyClientExecutor property (same for all instances, but thats OK). ProxyFunctionDescriptorProvider then no longer takes that as a ctor param. ScriptHost just passes in the metadatas.

This way, all the proxy special casing can go away in ScriptHost.GetFunctionsMetadata

ankitkumarr replied:
I see what you mean. Ah, I just realized that the `FunctionMetada` has `IsProxy` as well. Part of my concern was, if we remove the proxy metadata, how do we know if we need to add proxy descriptor or not. It can check if `FunctionMetada.IsProxy` is true for any.

And then you are saying that `ProxyFunctionDescriptorProvider` would check if the `FunctionMetada` is of `ProxyFunctionMetadata` type, if so, it will get the executor from there and use it? If not, that's an error anyway.

Does that sound accurate?

mathewc replied:
Right

mathewc commented on src/WebJobs.Script.Abstractions/Description/IFunctionProvider.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace Microsoft.Azure.WebJobs.Script.Abstractions
+{
+    public interface IFunctionProvider
```
This is basically the same as the existing IFunctionMetadataProvider interface we use internally. For the new public type I think that's the name we want to use. Our existing FunctionMetadataProvider type should then implement this public interface - it becomes a "composite provider" that aggregates all the metadata produced by registered providers. We just need to work out how to load all the custom providers from DI, then replace those with our single composite provider. Perhaps @Fabio has an idea on that.

We can remove the "forceRefresh" caching behavior from this interface, and have it always load fresh, and handle caching upstream where it should be handled, in FunctionMetadataManager which is the primary consumer of this.

ankitkumarr replied:
Or, what if we make `FunctionMetadataProvider` the `DefaultFunctionMetadaProvider` and not a composite provider? The `FunctionMetadaManager` can act as a composite which would take all of the `IFunctionMetadataProvider`s including the `DefaultFunctionMetadaProvider`, any custom ones, and create the collection?
Similar to what I have, but still making it consistent with our internal implementation.

That would mean we remove `IFunctionProvider` and move `IFunctionMetadataProvider` to abstractions and use that for all this, as you suggested.

mathewc replied:
That's a good idea - FunctionMetadataManager can be responsible for aggregating/returning metadata from all registered providers.

ankitkumarr replied:
I discussed this with @fabiocav, and also looked into the design. I think it makes sense to keep the `FunctionMetadataProvider` and custom providers separate.

- One benefit it provides is -- we can keep `bool forcerefresh` only limited to our internal type. We need this force refresh in Provider and not the Manager, as FunctionMetadataManager refreshes its list of providers any time script Host is restarted to account for any changes in such services. By having forcerefresh, we avoid having the provider relook at the filesystem, even though FunctionMetadataManager reiterates over all providers.

- Another key difference is -- `FunctionMetadataProvider` is WebHost scoped. Whereas, these `IFunctionProviders` providers (including proxy) are JobHost scoped and brought it later. They may change over the lifetime of the WebHost.

- I also like the separation of providers with "Custom" or "Virtual" functions, vs actual metadata from filesystem.

To keep such separation, `IFunctionMetadataManager` still acts as a composite, but looks at a single `IFunctionMetadataProvider` and enumerates over all registered `IFunctionProvider` to get the collection of the metadata.

I hope that this is reasonable. We can also discuss this in a different communication channel too, if you'd like.

mathewc commented on src/WebJobs.Script/Host/ScriptHost.cs:
```diff
@@ -524,6 +524,18 @@ private void AddFunctionDescriptors()
+        /// </summary>
+        private void AddCodelessDescriptor()
+        {
+            if (_functionMetadataManager.Functions.Any(m => string.Equals(m?.Language, DotNetScriptTypes.Codeless, StringComparison.OrdinalIgnoreCase)))
```
We should remove the special handling in this file for IProxyMetadataManager and treat it the same as these custom providers. If we register ProxyMetadataManager as an IFunctionMetadataProvider with DI that should happen naturally, and we can register ProxyFunctionDescriptorProvider in AddFunctionDescriptors as well, where it should be. Not sure why we're doing that in GetFunctionsMetadata right now

mathewc commented on test/WebJobs.Script.Tests/FunctionMetadataManagerTests.cs:
```diff
@@ -6,6 +6,7 @@
 using System.Collections.Immutable;
 using System.Collections.ObjectModel;
 using System.IO;
+using Microsoft.Azure.WebJobs.Script.Abstractions;
```
We have E2E tests for proxies (ProxyEndToEndTests) that verify that ListFunctions etc. surface proxy "virtual" functions. We should have an E2E test suite for codeless as well that tests the whole scenario/pipeline there as well.

ankitkumarr replied:
Good point, thanks!
I will add the recommended tests, once we are good with the design, just to avoid throw-away tests, especially if the interface changes a lot. I will leave this unresolved for now, will update before checking in.

ankitkumarr commented on src/WebJobs.Script/Workers/Rpc/RpcWorkerChannel.cs:
```diff
@@ -174,12 +174,19 @@ internal void WorkerInitResponse(RpcEvent initEvent)
             {
-                _workerChannelLogger.LogDebug("Setting up FunctionInvocationBuffer for function:{functionName} with functionId:{id}", metadata.Name, metadata.FunctionId);
-                _functionInputBuffers[metadata.FunctionId] = new BufferBlock<ScriptInvocationContext>();
+                if (Utility.IsFunctionMetadataLanguageSupportedByWorkerRuntime(metadata, _workerConfig?.Description?.Language))
```
@pragnagopa pointed out offline that this should be filtered upstream (the caller) and not in this channel here,. I will address that as well.

ankitkumarr commented on test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/SamplesEndToEndTests_CSharp.cs:
```diff
@@ -480,7 +480,7 @@ public async Task ListFunctions_Succeeds()
             var metadata = (await response.Content.ReadAsAsync<IEnumerable<FunctionMetadataResponse>>()).ToArray();
 
-            Assert.Equal(16, metadata.Length);
+            Assert.Equal(7, metadata.Length);
```
@fabiocav @mathewc, this PR makes it so that "List Functions" honors functions whitelist, and doesn't return non-whitelisted functions. It does make sense. Had a brief chat with @mathewc offline.

Thoughts?

ankitkumarr replied:
Fabio pointed out that we do document `Functions` in host.json-- https://docs.microsoft.com/en-us/azure/azure-functions/functions-host-json#functions

So, to avoid a breaking change here, I am going to keep the behavior as is -- include all functions in list functions call. I will push an update shortly.

ankitkumarr replied:
I updated this to not change anymore.

It seems that "Sync Triggers" call also indeed ignores the `Functions` property. Though, with this change that "Sync Triggers" will start honoring this. And I think it makes sense. As there's no reason to monitor those functions.

ankitkumarr commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -18,36 +21,81 @@ namespace Microsoft.Azure.WebJobs.Script
+
+            // Resetting the logger switches the logger scope to Script Host level,
+            // also making the logs available to Application Insights
+            _logger = _serviceProvider?.GetService<ILoggerFactory>().CreateLogger(LogCategories.Startup);
```
I want to explicitly call this out, to be sure it makes sense and is ok.

mathewc replied:
Not sure I understand this "reset services" concept

ankitkumarr replied:
FunctionMetadataManager is created now at the scope of the WebHost. This means that when it is initialized, the `IFunctionProvider`s would not be available yet. These providers come from the extensions controlled by the user code. So even though `FunctionMetadataManager` is at WebHost level, we need to "reset" these services everytime JobHost starts. This way we make sure any code changes reflect the proper metadata.

ankitkumarr commented:
I updated the PR to address all concerns and added some new tests. @mathewc, this should be good for another review.
I have also updated the namespace for the exposed types in my last commit.

fabiocav commented on src/WebJobs.Script.WebHost/Management/WebFunctionsManager.cs:
```diff
@@ -59,22 +59,18 @@ internal static async Task<IEnumerable<FunctionMetadataResponse>> GetFunctionMet
 
-            return functionsMetadata;
+            // Currently we don't want to return codeless functions in any web based APIs
+            return Utility.FilterOutCodeless(functionsMetadata);
```
mostly a nit: This would probably be more natural as a predicate we can pass to filter the collection. We can also combine the predicates here so we don't have to iterate over the collection multiple times

ankitkumarr replied:
With the predicate approach, there's no need for Utility functions. I just call the `.Where()`. Thanks. I updated to use a predicate here.

fabiocav commented on src/WebJobs.Script/Description/DotNet/DependencyHelper.cs:
```diff
@@ -165,5 +166,33 @@ public static List<string> GetRuntimeFallbacks(string rid)
+        /// </summary>
+        /// <param name="assemblyFormatString"> The assembly name string in the expected format. </param>
+        /// <returns> bool if the string was in the shared assembly format. </returns>
+        public static bool TryGetSharedAssembly(string assemblyFormatString, out string sharedAssembly)
```
The concept of shared assemblies already exists, and this overloads that. We should use a different name here. Assembly reference might fit better.

fabiocav commented on src/WebJobs.Script/Description/DotNet/DynamicFunctionAssemblyLoadContext.cs:
```diff
@@ -6,6 +6,7 @@
 using System.Reflection;
 using System.Runtime.Loader;
 using System.Text;
+using Microsoft.Azure.WebJobs.Script.Abstractions.Description;
```
I know @mathewc was in favor the namespace renames, but I'm wondering if that's something we should really do. The impact was quite significant and many of these types were public.

ankitkumarr replied:
@fabiocav, other than core-tools do you think anyone else depends on it? Though, I realize this is public and we can't control or know of who uses it.

mathewc replied:
Well, even moving the Types to a different assembly is a breaking change, so we've already decided on that right (though we could have added type forwarding)? I'm not aware of anyone consuming these types but us, but I see your point.

fabiocav replied:
Missed updating this, but adding `Abstractions` to the namespace here doesn't make a lot of sense and doesn't really follow the convention set in other projects. These are just grouped in the same package/assembly, but should follow the appropriate namespace for the area.

fabiocav commented on src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs:
```diff
@@ -78,6 +79,14 @@ public sealed class DotNetFunctionInvoker : FunctionInvokerBase
+            // as all shared assembly should already be resolved.
+            if (DependencyHelper.IsSharedAssemblyFormat(functionMetadata.ScriptFile))
+            {
+                return new NoOpFunctionMetadataResolver();
```
Return a singleton here. No need to create a new instance for each.

fabiocav commented on src/WebJobs.Script/Description/DotNet/NoOpFunctionMetadataResolver.cs:
```diff
@@ -0,0 +1,64 @@
+
+        private static Task<PackageRestoreResult> GetEmptyPackageResult()
+        {
+            return Task.FromResult(new PackageRestoreResult()
```
This task can also be cached.

ankitkumarr replied:
This private method is only ever called once, as it's Lazily called by `_emptyPackageResult`. So, I am not sure if there's any more need to cache this.

fabiocav commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -18,36 +21,95 @@ namespace Microsoft.Azure.WebJobs.Script
+        public FunctionMetadataManager(IOptions<ScriptJobHostOptions> scriptOptions, IFunctionMetadataProvider functionMetadataProvider, IEnumerable<IFunctionProvider> functionProviders, IOptions<HttpWorkerOptions> httpWorkerOptions, IScriptHostManager scriptHostManager, ILoggerFactory loggerFactory)
         {
             _scriptOptions = scriptOptions;
+            _serviceProvider = scriptHostManager as IServiceProvider;
```
Can you rename `_serviceProvider` here to something more descriptive like `_jobHostServiceProvider` or something along those lines?

fabiocav commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -18,36 +21,95 @@ namespace Microsoft.Azure.WebJobs.Script
+            }
+
+            // Add metadata and errors from any additional function providers
+            if (_functionProviders != null && _functionProviders.Any())
```
nit: `_functionProviders?.Any()`

ankitkumarr replied:
It'd have to be `_functionProviders?.Any() ?? false`. Not sure which is better. :)

mathewc replied:
I'd just remove this if block entirely, and compress your two Add methods into a single LoadCustomProviderFunctions method that does the null check and adds metadata/errors

fabiocav commented on src/WebJobs.Script/Host/IFunctionMetadataManager.cs:
```diff
@@ -2,14 +2,14 @@
         ImmutableDictionary<string, ImmutableArray<string>> Errors { get; }
 
-        ImmutableArray<FunctionMetadata> Functions { get; }
+        ImmutableArray<FunctionMetadata> GetFunctionMetadata(bool forceRefresh = false, bool includeBlocked = false);
```
If this also took predicate (Func over the metadata items), you'd be able to simplify the calls to the utility methods.

ankitkumarr replied:
With the other suggested, I think this may not be necessary as I ended up being able to remove Utility Functions altogether. And callers can call `.Where(Func<T,bool> predicate)` if they need to filter.

Not having it there, avoids having to re-create another ImmutableArray after filtering to fulfil the return type contract.

mathewc replied:
Here and other places we're also making breaking changes to public types, because we're pretty confident we're the only one using them

mathewc commented on src/WebJobs.Script.Abstractions/Description/FunctionMetadata.cs:
```diff
@@ -7,7 +7,7 @@
 using System.Linq;
 
-namespace Microsoft.Azure.WebJobs.Script.Description
+namespace Microsoft.Azure.WebJobs.Script.Abstractions.Description
```
Before we release the new package with all these new types, I think we should do a final pass on the public surface area. These types were never meant for public consumption, and there are lots of public helpers and other methods that I think we can move from the public surface area. E.g. FunctionMetadata.IsHttpInAndOutFunction, BindingMetadata.Create, etc. With these being elevated as a new extensibility model, we should review and only include what has to be there. We can do this in a separate PR - doesn't need to block this.

mathewc commented on src/WebJobs.Script/Extensions/FunctionMetadataExtensions.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace Microsoft.Azure.WebJobs.Script.Extensions
+{
+    public static class FunctionMetadataExtensions
```
When we do the pass over the public surface area of the new Types, some of the helpers like FunctionMetadata.IsHttpInAndOutFunction can move here as well I think

mathewc commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -18,36 +21,95 @@ namespace Microsoft.Azure.WebJobs.Script
+        /// Gets the function metadata array from all providers.
+        /// </summary>
+        /// <param name="forceRefresh">Forces reload from all providers.</param>
+        /// <param name="includeBlocked">Includes all functions, including blocked ones.</param>
```
What is a "blocked" function?

ankitkumarr replied:
If `Functions` in host.json is specified for the list of enabled functions, the functions that are not present there is what I termed as "blocked" function. Is there a better name?

pragnagopa replied:
Do you mean disabled functions?

mathewc replied:
I see. It goes to your earlier question on whether we should include non-whitelisted in this response. Since the whole point of whitelisting is to exclude functions completely, I'm wondering if we shouldn't just remove this switch and never return non-whitelisted functions. An "excluded" function won't be loaded by the host, won't have a listener, it won't be running, and it shouldn't be shown in portal, be monitored by ScaleController, etc. I know you had identified a discrepency where the admin APIs weren't applying the whitelist and you're trying to preserve that. Excluded functions also shouldn't be returned as part of the sync triggers response. Is it currently?

Perhaps we take this opportunity to clean that up. The whitelist is rarely used by people, so likely even before your changes we're listing functions that we shouldn't be. I'm leaning towards removing this flag and applying the whitelist consistently.

ankitkumarr replied:
@pragnagopa, these are for non-whitelisted functions.

Yes, today it is in fact returned as part of the sync triggers payload. Right now, I removed it from sync triggers payload but left it in "List Functions" to avoid breaking change. But with my recent understanding that "List Functions" uses the sync triggers cache, that will create a discrepancy.

@fabiocav mentioned earlier that we do have this `functions` (whitelist) property in the `host.json` schema as well as our documentation https://docs.microsoft.com/en-us/azure/azure-functions/functions-host-json#functions

Though, reading it, we specifically ask users to not do that in Azure.

@fabiocav had some concerns with breaking this behavior. I would love to hear his thoughts here too.

mathewc replied:
Rename this param to applyWhitelist to match existing terminology

mathewc commented on src/WebJobs.Script.Abstractions/Description/IFunctionProvider.cs:
```diff
@@ -0,0 +1,23 @@
+        /// Gets any function errors that may occur as part of the provider context
+        /// </summary>
+        /// <returns> An ImmutableDictionary of function name to the array of errors</returns>
+        ImmutableDictionary<string, ImmutableArray<string>> FunctionErrors { get; }
```
Why here and below do these have to be ImmutableArrays rather than just IEnumerables? IEnumerable is more natural.

ankitkumarr replied:
hmm, yeah initially I tried to keep this same as `IFunctionMetadataProvider` which uses immutable. I guess just so if it's indexed, it should never ever be changed.
But I don't have a strong preference here.

Changing the contract later would be a breaking change. So yeah good point, we should be certain about the types.

Because it's reconstructed into a compiled list anyway in MetadataManger. I suppose IEnumerable would work fine.

mathewc commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -89,5 +153,32 @@ internal bool IsScriptFileDetermined(FunctionMetadata functionMetadata)
+
+            var functionMetadataListArray = Task.WhenAll(functionProviderTasks).GetAwaiter().GetResult();
+
+            foreach (var someArray in functionMetadataListArray)
```
nit: rename someArray to currArray or something

mathewc commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -89,5 +153,32 @@ internal bool IsScriptFileDetermined(FunctionMetadata functionMetadata)
+
+            foreach (var someArray in functionMetadataListArray)
+            {
+                if (!someArray.IsDefaultOrEmpty)
```
Can you remove this DefaultOrEmpty check, and the ToList/ForEach below and just call functionMetadataList.AddRange?

ankitkumarr replied:
With recent changes, we can't do `AddRange` as we need to look at each metadata individually. So with the type as `Immutable`, we'd have to continue this way.

mathewc commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -89,5 +153,32 @@ internal bool IsScriptFileDetermined(FunctionMetadata functionMetadata)
+        {
+            foreach (var provider in _functionProviders)
+            {
+                provider.FunctionErrors?.ToList().ForEach(kvp => _functionErrors[kvp.Key] = kvp.Value.ToList());
```
Seems we need to have a check when aggregating functions from custom providers to prevent multiple functions of the same name from different providers?

ankitkumarr replied:
Yeah good point. I figured that would be handled upstream, but this should be handled here. I added logic to deal with that, and added tests for it. Thanks!

mathewc commented on src/WebJobs.Script/Host/ProxyFunctionMetadata.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace Microsoft.Azure.WebJobs.Script
+{
+    public class ProxyFunctionMetadata : FunctionMetadata
```
With this new Type, we should remove the FunctionMetadata.IsProxy property from the public surface area

ankitkumarr replied:
I think it may still help when we filter proxy functions in some places. I can replace it with `if (metadata is ProxyFunctionMetadata)` which adds a slight more overhead (I think?).

mathewc commented on src/WebJobs.Script/Host/ProxyFunctionProvider.cs:
```diff
@@ -20,43 +22,43 @@
 namespace Microsoft.Azure.WebJobs.Script
 {
-    public class ProxyMetadataManager : IProxyMetadataManager, IDisposable
+    public class ProxyFunctionProvider : IFunctionProvider, IDisposable
```
Nice

mathewc commented on src/WebJobs.Script/Host/ScriptHost.cs:
```diff
@@ -335,26 +335,12 @@ private void LogHostFunctionErrors()
             }
 
-            // Get proxies metadata
-            var proxyMetadata = _proxyMetadataManager.ProxyMetadata;
```
Nice

mathewc commented on src/WebJobs.Script/Host/ScriptHost.cs:
```diff
@@ -502,8 +488,14 @@ internal async Task InitializeFunctionDescriptorsAsync(IEnumerable<FunctionMetad
+        private void AddFunctionDescriptors(IEnumerable<FunctionMetadata> functionMetadata)
         {
+            // Add the proxy descriptor provider if needed
+            if (functionMetadata.Any(m => m?.IsProxy ?? false))
```
There won't (shouldn't) be any nulls in this collection. Should ensure that elsewhere. So this check should just be Any(p => p.IsProxy)

mathewc commented on src/WebJobs.Script/Host/ScriptHost.cs:
```diff
@@ -524,6 +516,18 @@ private void AddFunctionDescriptors()
+        /// </summary>
+        private void AddCodelessDescriptor()
+        {
+            if (_functionMetadataManager.GetFunctionMetadata().Any(m => string.Equals(m?.Language, DotNetScriptTypes.Codeless, StringComparison.OrdinalIgnoreCase)))
```
remove the null checks on metadata here and elsewhere. Make sure no nulls get into the collection when its built

mathewc commented on src/WebJobs.Script/ScriptHostBuilderExtensions.cs:
```diff
@@ -233,8 +234,8 @@ public static void AddCommonServices(IServiceCollection services)
             var logger = loggerFactory?.CreateLogger<ScriptStartupTypeLocator>() ?? throw new ArgumentNullException(nameof(loggerFactory));
-            var metadataServiceProvider = applicationHostOptions.RootServiceProvider.GetService<IFunctionMetadataProvider>();
-            return builder.UseExternalStartup(new ScriptStartupTypeLocator(applicationHostOptions.ScriptPath, logger, extensionBundleManager, metadataServiceProvider, metricsLogger), loggerFactory);
+            var metadataServiceManager = applicationHostOptions.RootServiceProvider.GetService<IFunctionMetadataManager>();
```
nit: functionMetadataManager

mathewc commented on src/WebJobs.Script/Utility.cs:
```diff
@@ -626,6 +626,11 @@ public static bool IsSupportedRuntime(string workerRuntime, IEnumerable<RpcWorke
         private static bool ContainsFunctionWithWorkerRuntime(IEnumerable<FunctionMetadata> functions, string workerRuntime)
         {
+            // Codeless is allowed for all worker runtimes
+            if (functions.Any(f => string.Equals(f.Language, DotNetScriptTypes.Codeless, StringComparison.OrdinalIgnoreCase)))
```
This doesn't seem right. This is called from Utility.IsSingleLanguage to ensure no mixed languages. If I have mixed Node/.NET and also a codeless, won't this short circuit the below required checks?

ankitkumarr replied:
This function checks if there's any function in the list that is acceptable for the current WorkerRuntime. So, say if you have `.NET` and `Node` functions, with `Node` as the worker rutime, it returns true as is. The `IsSingleLanguage` seems to check if it's a single language or if there is at least one function that matches the workerRuntime set.

In that sense, we are defining that `Codeless` is acceptable for all language workers. So, if you have `.NET`, `Node` and `Codeless` functions with worker runtime set to `Node`, that is allowed.
Additionally, if you have Worker Runtime set to `node` and you only have Codeless and .NET functions, that's fine too. Because Codeless is an acceptable function for Node worker runtime.

mathewc replied:
I find the IsSingleLanguage check confusing. Only if the WorkerRuntime isn't set does it appear to verify that functions declare a single language. If it is set (the normal case), it returns true if there is at least one function for the specified language, but doesn't ensure there is a single, which doesn't match the behavior described by the method name. Your check seems to short circuit this - it'll always return true. @pragnagopa understands this logic best I believe since I think she wrote it. If she thinks this is correct, then OK :)

pragnagopa replied:
If FUNCTIONS_WORKER_RUNTIME is set, it takes precedence and only need to check if there atleast one function matching the worker runtime set. If worker runtime is not set, the check enforces all the functions have same language.  Rename 'IsSingleLanguage` to 'IsSingleLanguageOrContainsWorkerRuntimeFunctions` ?

Do update tests to include cases for CodeLess functions

ankitkumarr replied:
Yes, good point. I will add tests. I can change the name to `IsSingleLanguageOrContainsWorkerRuntimeFunctions`, but is that a bit verbose?

I worry about tests that are `IsSingleLanguage_FunctionsList_Null_FunctionsWorkerRuntime_Throws_ArgumentNullException` will become `IsSingleLanguageOrContainsWorkerRuntimeFunctions_FunctionsList_Null_FunctionsWorkerRuntime_Throws_ArgumentNullException`, :P unless I don't rename them.

Could we reduce the scope of this function to only check for if there's a single function language? Then let the caller call the other `ContainsFunctionWithWorkerRuntime` if needed?
Looks like only one caller `Utility.VerifyFunctionsMatchSpecifiedLanguage` actually passes a `workerRuntime`, and that can manually call `ContainsFunctionWithWorkerRuntime`. The other caller just sets `workerRuntime` to `null` and would never call `ContainsFunctionWithWorkerRuntime` anyway.

pragnagopa commented on src/WebJobs.Script/Description/Compilation/ICompilationService.cs:
```diff
@@ -4,6 +4,7 @@
 using System;
 using System.Collections.Generic;
 using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Script.Abstractions.Description;
```
Nit: New using statements. Needed?

ankitkumarr replied:
Yes, because `FunctionMetadata` was in `Microsoft.Azure.WebJobs.Script.Description` and moved to `Microsoft.Azure.WebJobs.Script.Abstractions.Description`

pragnagopa commented on src/WebJobs.Script/Host/ScriptHost.cs:
```diff
@@ -524,6 +516,18 @@ private void AddFunctionDescriptors()
+        }
+
+        /// <summary>
+        /// Adds a DotNetFunctionDescriptorProvider to the list of descriptors if any function metadata has language set to "codeless" in it.
```
Can you please explain a little more about the specific purpose of codeless? invokes and descirptors are specific to  DotNet? Will this ever extend to non-dot languages?

ankitkumarr replied:
Discussed offline.

pragnagopa commented on test/WebJobs.Script.Tests/UtilityTests.cs:
```diff
@@ -11,6 +11,7 @@
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.Azure.WebJobs.Logging;
```
Update tests for IsSingleLanguage

mathewc commented on src/WebJobs.Script/Description/DotNet/DotNetScriptTypes.cs:
```diff
@@ -5,6 +5,7 @@ namespace Microsoft.Azure.WebJobs.Script.Description
 {
     public class DotNetScriptTypes
     {
+        public const string Codeless = "Codeless";
```
I'm thinking we should be able to do away with this new "language" type. For the LA scenario, the language type should just be DotNetAssembly.  If we need to continue to discriminate internally functions that we loaded from custom providers, rather than use language type, we could just have an **internal** property on FunctionMetadata IsCodeless that we set for functions we get from custom providers. Likely for all custom providers we've loaded from 3rd parties, we want to validate that they are of type DotNetAssembly for now? 

If we need to conditionally list/hide these custom functions, we can do that based on the internal property. Or if we wanted to allow flexibility, we could have another public property IsListed or something to govern whether they should be returned when listing functions. The proxy provider would then also set this to False.

ankitkumarr replied:
Discussed offline, and updated with changes that addresses some of these concerns.

mathewc commented on src/WebJobs.Script.Abstractions/Description/FunctionMetadata.cs:
```diff
@@ -7,7 +7,7 @@
-namespace Microsoft.Azure.WebJobs.Script.Description
+namespace Microsoft.Azure.WebJobs.Script.Abstractions.Description
 {
     public class FunctionMetadata
```
We should be able to remove IsProxy from this type, and use an internal subclass for proxies. Then everywhere we check IsProxy currently, we just do a type check. With this being a public type 3rd parties are coding against now, we don't want them to have to think about this. Same goes for IsDirect and IsHttpInAndOutFunction.

Review the other public members as well. E.g. IsDisabled, FunctionId, Trigger should also be removed. We need to find a way to continue to ferry that info around, without those members being part of the public type. E.g. the IsDisabled programming model for v3 is app settings based - we don't want the metadata provider specifying this.

mathewc commented on src/WebJobs.Script.WebHost/Extensions/FunctionMetadataExtensions.cs:
```diff
@@ -40,7 +40,7 @@ public static async Task<FunctionMetadataResponse> ToFunctionMetadataResponse(th
                 IsDirect = functionMetadata.IsDirect,
                 IsDisabled = functionMetadata.IsDisabled,
-                IsProxy = functionMetadata.IsProxy,
+                IsProxy = functionMetadata is ProxyFunctionMetadata,
```
Since we'll want IsDirect and other properties to also go away, but we still need to be able to associate our own internal info with these instances for tracking. One way would be to add a "Dictionary<string, string> Tags" member to the public interface that can be used to track these types of things, and we'd have a set of extension methods for IsDirect, IsProxy, etc. that read from this.

It's either that, or we define our own subclass that we copy provider instances into, that has the properties we need. However, that'd mean we'll be flowing this subclass type around rather than base FunctionMetadata.The benefit here is that it's strongly typed.

mathewc commented on src/WebJobs.Script.WebHost/Management/WebFunctionsManager.cs:
```diff
@@ -59,22 +60,22 @@ internal static async Task<IEnumerable<FunctionMetadataResponse>> GetFunctionMet
-                {
-                    functionsMetadata = proxyFunctionsMetadata.Concat(functionsMetadata);
-                }
+                // Remove all codeless metadata (includes proxies)
```
So we're saying that all Codeless functions are always excluded when listing functions. We should include that detail in the public comment for the FunctionMetadata member.

ankitkumarr replied:
Yes, I did add it in the `<remarks></remarks>

https://github.com/Azure/azure-functions-host/pull/5657/files/9bfde909aff43c7c6e40f2a2e2dfc41ff3a0ba37#diff-7a0d8f754916c5e5775de01cbf573b4eR58

mathewc commented on src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs:
```diff
@@ -359,7 +360,7 @@ public async Task<SyncTriggersPayload> GetSyncTriggersPayload()
         {
             var triggers = (await functionsMetadata
-                .Where(f => !f.IsProxy)
+                .Where(f => !(f is ProxyFunctionMetadata))
```
Might be good to add an IsProxy extension method to FunctionMetadataExtensions to do this check

ankitkumarr replied:
Yeah makes sense. I did that, then removed it when I added generic codeless, but I am still doing this check enough times to make this a useful extension.

mathewc commented on src/WebJobs.Script/Description/DotNet/NoOpFunctionMetadataResolver.cs:
```diff
@@ -0,0 +1,79 @@
+        {
+        }
+
+        public static NoOpFunctionMetadataResolver GetInstance()
```
Can you follow the pattern we use other places for this, like NullLinuxContainerActivityPublisher?

mathewc commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -18,36 +22,91 @@ namespace Microsoft.Azure.WebJobs.Script
+            return GetLoadedMetadata(includeBlocked);
+        }
+
+        internal ImmutableArray<FunctionMetadata> GetLoadedMetadata(bool includeBlocked)
```
Let's remove this GetLoadedMetadata function as well as the two _functionMetadataAllArray/_functionMetadataFilteredArray members, and just do the filtering herebefore returning. It's just a simple filters over a small collection - not a perf issue, given this API isn't called on hot paths. To handle your includeBlocked param, just do the whitelist filter here. Rename the param to applyWhitelist = true - I'd rather not have a new "blocked" alias with the existing "whitelist" concept.

mathewc commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -18,36 +22,91 @@ namespace Microsoft.Azure.WebJobs.Script
+            _functionProviders = functionProviders;
 
-        public ImmutableArray<FunctionMetadata> Functions => _functionMetadataArray.Value;
+            scriptHostManager.HostInitializing += ResetScriptHostServices;
```
Don't we need to wait until Initialized to ensure all extensions are loaded? I can't remember the event ordering.

mathewc replied:
Also, perhaps rather than adding this new event to the interface and subscribing, you can instead have ScriptHost call IFunctionMetadataManager.Reset at the right time?

ankitkumarr replied:
This is just subscribing to the `ScriptHostInitializing`. When Script Host is initializing, all extensions have already been loaded. SciptHost invokes this right before getting new metadata -- https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/Host/ScriptHost.cs#L276-L279

Having this be an event handler vs making ScriptHost call `IFunctionMetadataManager.Reset` avoids ScriptHost having to worry about the implementation detail. Also, it doesn't seem right to expose this in the interface for `IFunctionMetadataManager`.
It makes sense to me that the ScriptHost opens up a hook, that other callers can subscribe to.
I think this is also something that I had some discussion with @fabiocav at early stages.

mathewc commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -18,36 +22,91 @@ namespace Microsoft.Azure.WebJobs.Script
+            return includeBlocked ? _functionMetadataAllArray : _functionMetadataFilteredArray;
+        }
+
+        private void ResetScriptHostServices(object sender, EventArgs e)
```
Rename this to just Reset, and call this from the ctor to do the init. That way we have a single place we're maintaining this.

ankitkumarr replied:
I had it `Reset` initially, but Fabio made a good point that we should be clearer and explicitly call out in the name that we are resetting ScripHost level services, and suggested the name `ResetScriptHostServices`. I think I like that it's clear specially if it's only a private function and can call out specifics.

ankitkumarr replied:
Also calling initialize in ctor is a bit tricky because the Script Host level service provider doesn't yet have an logging initialized and I get a null exception. So, until Script Host is initializing, we can't start getting services from it's service provider.

Maybe I should add comments about that, the fact that these services are changing scopes,

mathewc commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -89,5 +150,56 @@ internal bool IsScriptFileDetermined(FunctionMetadata functionMetadata)
             return true;
         }
+
+        private void LoadCustomProvidersFunctions(List<FunctionMetadata> functionMetadataList)
```
LoadCustomProviderFunctions

mathewc commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -89,5 +150,56 @@ internal bool IsScriptFileDetermined(FunctionMetadata functionMetadata)
+            }
+        }
+
+        private void AddErrorsFromCustomProviders()
```
just remove this function and make it part of the function above,

mathewc commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -89,5 +150,56 @@ internal bool IsScriptFileDetermined(FunctionMetadata functionMetadata)
+        {
+            foreach (var provider in _functionProviders)
+            {
+                provider.FunctionErrors?.ToList().ForEach(kvp => _functionErrors[kvp.Key] = kvp.Value.ToList());
```
nit: just do a foreach (var error in provider.FunctionErrors), so you don't need the ToList()

mathewc commented on src/WebJobs.Script/Host/ScriptHost.cs:
```diff
@@ -303,6 +301,8 @@ public async Task InitializeAsync(CancellationToken cancellationToken = default)
                 await InitializeFunctionDescriptorsAsync(functionMetadataList);
 
                 // Initialize worker function invocation dispatcher only for valid functions after creating function descriptors
+                // We do not need any dipatcher for all codeless functions except Proxies.
```
nit: "Dispatcher not needed for non-proxy codeless functions"?

mathewc commented on src/WebJobs.Script/Host/ScriptHost.cs:
```diff
@@ -530,6 +516,30 @@ private void AddFunctionDescriptors()
+            // If we have a non-proxy codeless function, we need to add a .NET descriptor provider. But only if it wasn't already added.
+            // At the moment, we are assuming that all codeless functions will have language as DotNetAssembly
+            if (functionMetadata.Any(m => m.IsCodeless && !(m is ProxyFunctionMetadata))
+                && !_descriptorProviders.Any(d => d is DotNetFunctionDescriptorProvider))
```
nit: make the DotNetFunctionDescriptorProvider check first to avoid work in some cases

mathewc commented on src/WebJobs.Script/Utility.cs:
```diff
@@ -569,23 +569,23 @@ internal static bool IsSingleLanguage(IEnumerable<FunctionMetadata> functions, s
             }
-            var functionsListWithoutProxies = functions.Where(f => f.IsProxy == false).ToArray();
-            if (functionsListWithoutProxies.Length == 0)
+            var functionsListWithoutProxiesCodeless = functions.Where(f => !f.IsCodeless).ToArray();
```
nit: rename to filteredFunctions

mathewc commented on src/WebJobs.Script/Utility.cs:
```diff
@@ -569,23 +569,23 @@ internal static bool IsSingleLanguage(IEnumerable<FunctionMetadata> functions, s
             if (IsSingleLanguage(functions, null))
             {
-                var functionsListWithoutProxies = functions?.Where(f => f.IsProxy == false);
+                var functionsListWithoutProxies = functions?.Where(f => !(f is ProxyFunctionMetadata));
```
Rename. Also why do you filter out all codeless in the IsSingleLanguage call, but here only filter Proxies? Shouldn't the same filter be used both places? If so, move the filter above the IsSingleLanguage call and pass the filtered list in.

I'm not sure anyone is going to be able to understand this, and even you/I won't be able to a week from now :)

ankitkumarr replied:
I think filtering all codeless will be fine here. It's not as relevant.
I didn't really touch this code path, except change how we are checking for if something is proxy.

But yeah, you have a point and I can update this as well.
